### ISSUE TITLE
[State Timeline] Absence of data not represented anymore

### ISSUE THREAD

**Issue by sylr** [grafana/grafana#42868]
**What happened**:

I have a panel which shows the status of a node during its lifetime (either OK or OK).

```
abs(max(max_over_time(kube_node_status_condition{condition="Ready",status=~"(false|unknown)"}[1m:])) by (node) - 1)
```

![Screenshot 2021-12-08 at 10-47-25 k8s-nodes-thanos - Grafana](https://user-images.githubusercontent.com/153052/145186672-9c6bc22e-cd6b-4120-ae94-7c748438f7fc.png)

I pretty sure it used to work correctly and bars stopped when their corresponding data stopped having values because the nodes did not exist anymore. Now the bars continues till `now` even if there is no data.

**What you expected to happen**:

Bar should stop when there is no data instead of continuing with last `known value`.

**How to reproduce it (as minimally and precisely as possible)**:

<details>
<summary>Panel's JSON</summary>
<p>

```json
{
  "id": 106,
  "gridPos": {
    "h": 7,
    "w": 24,
    "x": 0,
    "y": 7
  },
  "type": "state-timeline",
  "title": "Availability periods",
  "repeatDirection": "h",
  "transformations": [
    {
      "id": "renameByRegex",
      "options": {
        "regex": "(.*).compute.internal",
        "renamePattern": "$1"
      }
    }
  ],
  "datasource": {
    "uid": "$datasource"
  },
  "description": "Incident period granularity is 1m",
  "links": [],
  "fieldConfig": {
    "defaults": {
      "custom": {
        "lineWidth": 4,
        "fillOpacity": 100
      },
      "color": {
        "mode": "continuous-GrYlRd"
      },
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "green",
            "value": null
          },
          {
            "color": "red",
            "value": 80
          }
        ]
      },
      "mappings": [
        {
          "options": {
            "0": {
              "color": "#bf1b00",
              "index": 0,
              "text": "KO"
            },
            "1": {
              "color": "#7EB26D",
              "index": 1,
              "text": "OK"
            }
          },
          "type": "value"
        }
      ],
      "unit": "short"
    },
    "overrides": []
  },
  "hideTimeOverride": false,
  "options": {
    "mergeValues": true,
    "showValue": "auto",
    "alignValue": "left",
    "rowHeight": 0.9,
    "legend": {
      "displayMode": "list",
      "placement": "bottom"
    },
    "tooltip": {
      "mode": "single"
    }
  },
  "targets": [
    {
      "exemplar": true,
      "expr": "abs(max(max_over_time(kube_node_status_condition{env=~\"$env\",k8s_cluster=~\"$k8scluster\",node=~\"$k8snode\",condition=\"Ready\",status=~\"(false|unknown)\"}[1m])) by (node) - 1)",
      "format": "time_series",
      "interval": "1m",
      "intervalFactor": 1,
      "legendFormat": "{{ node }}",
      "refId": "A"
    }
  ]
}
```

</p>
</details>

**Anything else we need to know?**:

**Environment**:
- Grafana version: [8.3.1+sylr.1](https://github.com/sylr/grafana/commits/v8.3.x%2Bsylr)
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

**Comment by ashharrison90** [comment]
hi @sylr 👋 

i can't reproduce this locally. we have a state timeline demo to explicitly test that gaps are present when the data is null. are you able to send the panel data also? panel header menu -> inspect -> data

thanks! 🙌

**Comment by ashharrison90** [comment]
@sylr was kind enough to share some data on slack. it looks like the presence of `undefined` is causing this.

i believe we made a change recently to this 🤔  you should still be able to force `undefined` to be blank by using a value mapping. can you try adding a value mapping for null/undefined, set the color to transparent and see if that helps?

**Comment by leeoniya** [comment]
> i believe we made a change recently to this 🤔

this has always been the design. `undefined` means continue/interpolate prior state and `null` means terminate prior state (explicit gap). it works the same as Time series.

i'm not aware of this ever working differently. 🤔

**Comment by sylr** [comment]
This might be my excel extract in "Series joined by time" that added undefined where it should be null. If I do an extract for a serie that only have a few values then nothing, I get only the few values, no undefined:

![image](https://user-images.githubusercontent.com/153052/145228856-49841221-5ede-41d0-9a3a-aa333cdd023f.png)
 
vs

![image](https://user-images.githubusercontent.com/153052/145229030-4c63ebd2-906b-429b-9eed-38c71a26f242.png)

**Comment by sylr** [comment]
I reverted to grafana 8.2.5 and it works as expected:

![image](https://user-images.githubusercontent.com/153052/145231102-e5dd476e-9186-4582-b382-f2a5eca2111b.png)

**Comment by argibbs** [comment]
I see this exact problem too, and have made the exact same fix (downgrading to 8.2.7)

FWIW I use both the state panel and the Nately discrete panel, and both have this problem.

**Comment by argibbs** [comment]
In case it helps, my source data is basically an int in [1,10] or \<nothing\>, e.g. here's a s/shot of the underlying prometheus query from the native prometheus viewer (series labels deliberately omitted, sorry).

I map these to different coloured bars in the status panel, or (prior to 8.3.0 at least) empty space when it's \<nothing\>

![image](https://user-images.githubusercontent.com/4406514/145307485-dcb104a5-e91d-482c-9701-7637184106d2.png)

**Comment by torkelo** [comment]
strange, before the transform I cannot see any null data (gaps). Can more share there raw data? and the data after join?

**Comment by sylr** [comment]
@torkelo how can you share the right way to extract data for you to investigate ?

**Comment by argibbs** [comment]
Here's a sanitised copy of the prometheus query that generates the above info

```
(clamp_max(sum(delta(metric1{warning="msg1"}[130s])) by (label1), 1) * 4) > 0
OR (clamp_max(sum(delta(metric2{result="generated"}[130s])) by (label1), 1) > 0
OR (clamp_max(sum(delta(metric1{warning="msg2"}[130s])) by (label1), 1) * 3) > 0
OR (clamp_max(sum(delta(metric3[130s])) by (label1), 1) * 2)) > 0
OR (clamp_max(sum(delta(metric2{result!="generated"}[130s])) by (label1), 1) * 5) > 0
```

I think it's overly complicated though; the main point is that I'm doing `>0` which causes some parts of the timeseries to be absent.

**Comment by sylr** [comment]
@torkelo ?

**Comment by ivanahuckova** [comment]
@sylr yes, you are correct. Thank you for connecting this issues. This is duplicate/same issue as https://github.com/grafana/grafana/issues/42993. I am going to close this so we have 1 issue where we track when is this fixed. https://github.com/grafana/grafana/issues/42993

**Comment by sylr** [comment]
@ivanahuckova I've built the tip of v8.3.x with your patch for #42993 but it does not resolve this particular issue.

Could you please re-open this issue ?

**Comment by ivanahuckova** [comment]
@sylr Could you please share the returned query response from the panel inspector from the version where it worked and where it stopped working? 
![image](https://user-images.githubusercontent.com/30407135/148236003-37db7417-2b6c-40d6-895a-b73d9964092b.png)

**Comment by ivanahuckova** [comment]
And also Data tab?

**Comment by sylr** [comment]
Fixed in v8.3.4.

**Comment by cykl** [comment]
@ivanahuckova Is that possible that 8.4.1 reintroduced this issue? I have not investigated it yet, but I just spotted a similar issue with 8.4.1 while the state timeline panel was fine with 8.3.4. Bar doesn't terminate when a time-series vanishes (basically my query is ALERT{alerstate="firing"})

**Comment by ivanahuckova** [comment]
I am not sure, but 8.4 has also some changes so it is possible. If you experience the issue, could you create a new issue for it? Thank you!

**Comment by hawran** [comment]
> Fixed in v8.3.4.

I am not sure, we are on `8.3.6` and we can see as follows:
![220303-112018 grafana state-timeline plugin last-30days](https://user-images.githubusercontent.com/25924718/156549964-d2991983-2a44-4772-8fbb-750ee9e18f65.png)
![220303-112242 grafana state-timeline plugin last-7days](https://user-images.githubusercontent.com/25924718/156549979-06200087-f6ba-4c7b-aa9b-e2b8812694b2.png)

NB We can not follow last versions in our production environment...

**Comment by leeoniya** [comment]
@hawran that was fixed in https://github.com/grafana/grafana/pull/45997. it will be in next release (8.4.4)

**Comment by hawran** [comment]
> @hawran that was fixed in #45997. it will be in next release (8.4.4)

Hi, na backport to `8.3.` stream?

**Comment by hawran** [comment]
> Hi there, this doesn't look to be fixed in 8.4.4 for me.

Got the same impression, trying to find out more details...
